src/st/cat/e_bone_ark.c

```diff
@@ -0,0 +1,1466 @@
+        dx += (siblingEntity->ext.boneArk.unk88.val << 8) >> 8;
+
+        angle = ((dx >> 0x10) * dx + (dy >> 0x10) * dy) >> 0x10;
+        if (angle > 0x24) {
+            angle = ratan2(dy, dx);
+            dx = rcos(angle) * 4 * 16;
+            dy = rsin(angle) * 4 * 16;
+        }
+        self->posX.val = siblingEntity->posX.val + dx;
+        self->posY.val = siblingEntity->posY.val - dy;
+        self->ext.boneArk.unk88.val = -(dx - xOffset);
+        self->ext.boneArk.unk8C.val = -(yOffset - dy);
+        break;
+    }
+}
+
+// Renders the flying pieces of ark which burst into flames
+static void RenderDeathParts(Primitive* prim) {
+    Entity* newEntity;
+
+    switch (prim->next->u2) {
+    case 0:
+        LOW(prim->next->u0) = (prim->next->r3 % 2) * 0x18000 - 0x8000 - 0x4000;
+        LOW(prim->next->u0) += ((Random() & 3) << 0xD) - 0x4000;
+        LOW(prim->next->r1) = -0x40000;
+        LOW(prim->next->r1) -= (prim->next->r3 >> 1) << 0xD;
+        LOW(prim->next->r1) += ((Random() & 3) << 0xE) - 0x8000;
+        prim->next->u2 = 1;
+        break;
+    case 1:
+        UnkPrimHelper(prim);
+        if (prim->next->r3 % 2) {
+            LOH(prim->next->tpage) += 0xD0 - ((prim->next->r3 >> 1) * 0x60);
+            prim->next->x3 += ((Random() & 3) * 0x10) + 0x100;
+            prim->next->y3 += ((Random() & 3) * 0x10) + 0x20;
+        } else {
+            LOH(prim->next->tpage) -= 0xD0 - ((prim->next->r3 >> 1) * 0x60);
+            prim->next->x3 -= ((Random() & 3) * 0x10) + 0x100;
+            prim->next->y3 -= ((Random() & 3) * 0x10) + 0x20;
+        }
+
+        LOW(prim->next->r1) += 0x3000 - ((prim->next->r3 >> 1) << 0xC);
+        if (LOW(prim->next->r1) > ((prim->next->r3 >> 1) << 0xF) + 0x34000) {
+            newEntity = AllocEntity(&g_Entities[224], &g_Entities[256]);
+            if (newEntity != NULL) {
+                CreateEntityFromCurrentEntity(E_EXPLOSION, newEntity);
+
+                if (prim->next->r3 >> 1) {
+                    newEntity->params = 2;
+                } else {
+                    newEntity->params = 3;
+                }
+
+                newEntity->facingLeft = prim->next->r3 % 2;
+                newEntity->posX.i.hi = prim->next->x1;
+                newEntity->posY.i.hi = prim->next->y0;
+            }
+            UnkPolyFunc0(prim);
+        }
+        break;
+    }
+}
+
+typedef enum BoneArkStep {
+    INIT = 0,
+    IDLE_WALK = 1,
+    CROUCHING = 2,
+    CHARGING = 3,
+    SHOT_FIRED = 4,
+    DEATH = 8
+};
+
+extern Primitive* FindFirstUnkPrim2(Primitive* prim, u8 index);
+void EntityBoneArk(Entity* self) {
+    s16 xOffset;
+    s16 yOffset;
+
+    Primitive* prim;
+    s32 i;
+    Entity* entity;
+    s16 var_s3;
+    s16 angle;
+    u8 var_s5;
+    s32 primIndex;
+    s32 dx;
+    s32 dy;
+    s32 pad[9];
+
+    FntPrint("main_step %x\n", self->step);
+    FntPrint("main_step_s %x\n", self->step_s);
+    FntPrint("hd_flag %x\n", self->ext.boneArk.headDying);
```

Comment:
I took a guess what `hd` might mean, this seems about as reasonable as anything I could think of 😄

Reply:
It's clearly the happy dance performed by the skeletons.  😆